### ISSUE TITLE
Bogus var on explain when ORDER BY contains aggregate

### ISSUE THREAD

**Issue by metdos** [citusdata/citus#756]
The query below works without a problem if you run it without explain. If you run it with explain we get this error:

```
ERROR:  bogus varattno for OUTER_VAR var: 3
```

Steps to replicate the problem:

On all nodes:

``` sql
CREATE TYPE user_composite_type AS
(
    tenant_id BIGINT,
    user_id BIGINT
);
```

On the master node:

``` sql
CREATE TABLE events (
    composite_id user_composite_type,
    event_id bigint,
    event_type character varying(255),
    event_time bigint
);
SELECT master_create_distributed_table('events', 'composite_id', 'range');

SELECT master_create_empty_shard('events') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(1,1)', shardmaxvalue = '(1,2000000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('events') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(1,2000000001)', shardmaxvalue = '(1,4300000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('events') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(2,1)', shardmaxvalue = '(2,2000000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('events') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(2,2000000001)', shardmaxvalue = '(2,4300000000)'
WHERE shardid = :new_shard_id;


CREATE TABLE users (
    composite_id user_composite_type,
    lastseen bigint
);
SELECT master_create_distributed_table('users', 'composite_id', 'range');

SELECT master_create_empty_shard('users') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(1,1)', shardmaxvalue = '(1,2000000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('users') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(1,2000000001)', shardmaxvalue = '(1,4300000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('users') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(2,1)', shardmaxvalue = '(2,2000000000)'
WHERE shardid = :new_shard_id;

SELECT master_create_empty_shard('users') AS new_shard_id
\gset
UPDATE pg_dist_shard SET shardminvalue = '(2,2000000001)', shardmaxvalue = '(2,4300000000)'
WHERE shardid = :new_shard_id;
```

And finally the failing explain query:

``` sql
SET citus.subquery_pushdown TO TRUE;

EXPLAIN SELECT
    avg(array_ndims(events)) AS dimension_avg,
    count_pay
    FROM (
  SELECT
    subquery_1.tenant_id,
    subquery_1.user_id,
    array_agg(event ORDER BY event_time) AS events,
    COALESCE(count_pay, 0) AS count_pay
  FROM
    (
        (SELECT
            (users.composite_id).tenant_id,
            (users.composite_id).user_id,
            'action=>1'AS event,
            events.event_time
        FROM
            users,
            events
        WHERE
            (users.composite_id).tenant_id = (events.composite_id).tenant_id AND
            (users.composite_id).user_id = (events.composite_id).user_id AND
            users.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
            users.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
            event_type = 'click')
        UNION
        (SELECT
            (users.composite_id).tenant_id,
            (users.composite_id).user_id,
            'action=>2'AS event,
            events.event_time
        FROM
            users,
            events
        WHERE
            (users.composite_id).tenant_id = (events.composite_id).tenant_id AND
            (users.composite_id).user_id = (events.composite_id).user_id AND
            users.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
            users.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
            event_type = 'submit')
    ) AS subquery_1
    LEFT JOIN
        (SELECT
            (composite_id).tenant_id,
            (composite_id).user_id,
            COUNT(*) AS count_pay
        FROM
            events
        WHERE
            events.composite_id >= '(1, -9223372036854775808)'::user_composite_type AND
            events.composite_id <= '(1, 9223372036854775807)'::user_composite_type AND
            event_type = 'pay'
        GROUP BY
            tenant_id,
            user_id
        HAVING
            COUNT(*) > 2) AS subquery_2
    ON
        subquery_1.tenant_id = subquery_2.tenant_id AND
        subquery_1.user_id = subquery_2.user_id
    GROUP BY
        subquery_1.tenant_id,
        subquery_1.user_id,
        count_pay) AS subquery_top
WHERE
    array_ndims(events) > 0
GROUP BY
    count_pay
ORDER BY
    count_pay;
```

**Comment by metdos** [comment]
Call stack:

```
errstart() at elog.c:245 0x10d7c70d2	
elog_finish() at elog.c:1,351 0x10d7cae3b	
get_variable() at ruleutils.c:5,924 0x10d74a7c2	
get_rule_expr() at ruleutils.c:7,060 0x10d745b46	
deparse_expression_pretty() at ruleutils.c:2,517 0x10d743f31	
deparse_expression() at ruleutils.c:2,475 0x10d743ea4	
show_sort_group_keys() at explain.c:2,026 0x10d3f764f	
show_sort_keys() at explain.c:1,826 0x10d3f63a8	
ExplainNode() at explain.c:1,476 0x10d3f38b5	
ExplainPrintPlan() at explain.c:585 0x10d3f1c21	
ExplainOnePlan() at explain.c:515 0x10d3f1775	
ExplainMasterPlan() at multi_explain.c:388 0x10de5eb3d	
MultiExplainOneQuery() at multi_explain.c:268 0x10de5e52f	
ExplainOneQuery() at explain.c:341 0x10d3f0f37	
ExplainQuery() at explain.c:245 0x10d3f0c2e	
standard_ProcessUtility() at utility.c:658 0x10d639c37	
```

**Comment by anarazel** [comment]
On 2016-08-29 06:53:00 -0700, Metin Döşlü wrote:
> Call stack:
> 
> ```
> errstart() at elog.c:245 0x10d7c70d2	
> elog_finish() at elog.c:1,351 0x10d7cae3b	
> get_variable() at ruleutils.c:5,924 0x10d74a7c2	
> get_rule_expr() at ruleutils.c:7,060 0x10d745b46	
> deparse_expression_pretty() at ruleutils.c:2,517 0x10d743f31	
> deparse_expression() at ruleutils.c:2,475 0x10d743ea4	
> show_sort_group_keys() at explain.c:2,026 0x10d3f764f	
> show_sort_keys() at explain.c:1,826 0x10d3f63a8	
> ExplainNode() at explain.c:1,476 0x10d3f38b5	
> ExplainPrintPlan() at explain.c:585 0x10d3f1c21	
> ExplainOnePlan() at explain.c:515 0x10d3f1775	
> ExplainMasterPlan() at multi_explain.c:388 0x10de5eb3d	
> MultiExplainOneQuery() at multi_explain.c:268 0x10de5e52f	
> ExplainOneQuery() at explain.c:341 0x10d3f0f37	
> ExplainQuery() at explain.c:245 0x10d3f0c2e	
> standard_ProcessUtility() at utility.c:658 0x10d639c37	
> ```

Given that is ExplainMasterPlan() at multi_explain.c:388 0x10de5eb3d:
		ExplainMasterPlan(masterPlan, into, es, queryString, params, &planDuration);

this suggest that the generated master query isn't correctly generated.

**Comment by metdos** [comment]
> Given that is ExplainMasterPlan() at multi_explain.c:388 0x10de5eb3d:
> 		ExplainMasterPlan(masterPlan, into, es, queryString, params, &planDuration);
> 
> this suggest that the generated master query isn't correctly generated.

I need to look into this more, but master query seems fine. I have checked MultiPlan->masterQuery before ExplainMasterPlan() is called, and it looks fine.

**Comment by metdos** [comment]
This is introduced with 5.1 where we added explain support. I replicated this with 5.1.0, so this is not a regression of afterward.

**Comment by onderkalaci** [comment]
We've recently hit the same issue on **non-subqueries** as well (so, changing the title now).

See the steps to re-produce:

```SQL

CREATE TABLE users_table (user_id int, time timestamp, value_1 int, value_2 int, value_3 float, value_4 bigint);
SELECT create_distributed_table('users_table', 'user_id');

EXPLAIN ANALYZE   Select user_id, max(time) from users_table a group by 1 order by avg(value_2) DESC;
ERROR:  bogus varattno for OUTER_VAR var: 4
```

Our initial investigation shows that some of the aggregates (i.e., `count` and `avg`) on the `ORDER BY` clause can lead to this problem.

**Comment by ozgune** [comment]
This issue could be related to #520 (Marco's note).

**Comment by mtuncer** [comment]
I also hit this when testing select distinct on() that contains aggregates and order by

**Comment by mtuncer** [comment]
minimal repro is
```sql
create table t1(a int, b int);
select create_distributed_table('t1', 'a');
explain select a, avg(b) from t1 group by a order by 1,2;
```

**Comment by saicitus** [comment]
+1 ran into this with a customer with a query below:
```
explain (analyze,buffers)
SELECT a.c1,a.c2,a.c3,
SUM(CASE WHEN b.c1 < 25 THEN 1 ELSE 0 END)/222 as lf,
SUM(CASE WHEN b.c1 >= 25 THEN 1 ELSE 0 END)/2282 as hf
FROM a_sharded a JOIN b_reference b ON (a.k = b.k)
GROUP BY a.c1,a.c2,a.c3 order by SUM(CASE WHEN b.c1 >= 25 THEN 1 ELSE 0 END)-SUM(CASE WHEN b.c1 < 25 THEN 1 ELSE 0 END);
```

**Comment by DimCitus** [comment]
My fix attempt still needs some help to review the explain plan changes in the regression tests. On a very quick pass through it @mtuncer seemed to think it was ok. This should be confirmed with a review of the PR, and that being the case, we could maybe fix this issue? See #2047.

**Comment by JelteF** [comment]
My changes on the last commit of this PR seem to fix this automatically as well: https://github.com/citusdata/citus/pull/3098
To see that it's fixed, search the test output for `varattno`:
https://circleci.com/gh/citusdata/citus/47900?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

(it currently breaks some other tests, but I'm in the process of fixing that)
